### PR TITLE
[RLlib] Fix SAC/DQN/CQL GPU and multi-GPU.

### PR DESCRIPTION
**Pull request by sven1977**
Fix DQN/SAC/CQL GPU and multi-GPU.
* Fix bug in DQN/SAC/CQL where the target net do not get updated due to missing timestep information on the Learner side (only happens when we have 1 or more remote Learners).
* For torch DDP to work with a complex setup like SAC's (where the same network (q-net) is passed twice, but one of these passes should NOT record gradients): Implement "straight-through" gradients for the q-net forward pass with the resampled actions (computed by the policy net). In other words, make sure that for this forward pass, the q-net does NOT get its gradients recorded, but the policy net does.
* Add CI learning tests for combinations of `[DQN | SAC] x [single-agent | multi-agent] x [CPU Learner | GPU Learner | 2 CPU Learners | 2 GPU Learners]`.
* Add release test for SAC on HalfCheetah-v4.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -349,6 +349,30 @@ py_test(
     srcs = ["tuned_examples/dqn/cartpole_dqn.py"],
     args = ["--as-test", "--enable-new-api-stack"]
 )
+py_test(
+    name = "learning_tests_cartpole_dqn_gpu",
+    main = "tuned_examples/dqn/cartpole_dqn.py",
+    tags = ["team:rllib", "exclusive", "learning_tests", "torch_only", "learning_tests_discrete", "learning_tests_pytorch_use_all_core", "gpu"],
+    size = "large",
+    srcs = ["tuned_examples/dqn/cartpole_dqn.py"],
+    args = ["--as-test", "--enable-new-api-stack", "--num-gpus=1"]
```
Comment (simonsays1980) [inline rllib/BUILD:358]: Does `num-gpus=1` use a local or remote learner? Imo, we should test with both. What do you think @sven1977 ?

File: rllib/BUILD [hunk 2]
```diff
@@ -358,16 +382,29 @@ py_test(
...
-    size = "large", # bazel may complain about it being too long sometimes - large is on purpose as some frameworks take longer
-    srcs = ["tests/run_regression_tests.py"],
-    data = ["tuned_examples/dqn/cartpole-dqn-softq.yaml"],
-    args = ["--dir=tuned_examples/dqn"]
+    name = "learning_tests_multi_agent_cartpole_dqn_gpu",
+    main = "tuned_examples/dqn/multi_agent_cartpole_dqn.py",
+    tags = ["team:rllib", "exclusive", "learning_tests", "torch_only", "learning_tests_discrete", "learning_tests_pytorch_use_all_core", "gpu"],
+    size = "large",
+    srcs = ["tuned_examples/dqn/multi_agent_cartpole_dqn.py"],
+    args = ["--as-test", "--enable-new-api-stack", "--num-agents=2", "--num-cpus=4", "--num-gpus=1"]
```
Comment (simonsays1980) [inline rllib/BUILD:391]: Interesting, I thought this does not work `--num-gpus > 0` and `--num-cpus > 0` :)

File: rllib/BUILD [hunk 3]
```diff
@@ -678,7 +739,22 @@ py_test(
...
+    size = "large",
+    srcs = ["tuned_examples/sac/multi_agent_pendulum_sac.py"],
+    args = ["--as-test", "--enable-new-api-stack", "--num-agents=2", "--num-cpus=4", "--num-gpus=1"]
+)
+py_test(
+    name = "learning_tests_multi_agent_pendulum_sac_multi_cpu",
+    main = "tuned_examples/sac/multi_agent_pendulum_sac.py",
+    tags = ["team:rllib", "exclusive", "learning_tests", "torch_only", "learning_tests_continuous"],
+    size = "large",
+    srcs = ["tuned_examples/sac/multi_agent_pendulum_sac.py"],
```
Comment (simonsays1980) [inline rllib/BUILD:755]: Do we actually need the `srcs` for files that can be executed directly via python?

File: rllib/algorithms/dqn/dqn.py [hunk 4]
```diff
@@ -628,13 +629,15 @@ def _training_step_new_api_stack(self, *, with_noise_reset) -> ResultDict:
...
                 )
-            # Add the sampled experiences to the replay buffer.
-            self.local_replay_buffer.add(episodes)
             # Reduce EnvRunner metrics over the n EnvRunners.
             self.metrics.merge_and_log_n_dicts(
                 env_runner_results, key=ENV_RUNNER_RESULTS
             )
 
+            # Add the sampled experiences to the replay buffer.
+            with self.metrics.log_time((TIMERS, REPLAY_BUFFER_ADD_DATA_TIMER)):
```
Comment (simonsays1980) [inline rllib/algorithms/dqn/dqn.py:638]: Nice :)

File: rllib/algorithms/sac/torch/sac_torch_rl_module.py [hunk 5]
```diff
@@ -103,9 +103,64 @@ def _forward_train(self, batch: Dict) -> Dict[str, Any]:
...
+            Columns.OBS: batch[Columns.OBS],
+            Columns.ACTIONS: actions_resampled,
+        }
+        # Make sure we perform a "straight-through gradient" pass here,
+        # ignoring the gradients of the q-net, however, still recording
+        # the gradients of the policy net (which was used to rsample the actions used
+        # here). This is different from doing `.detach()` or `with torch.no_grads()`,
+        # as these two methds would fully block all gradient recordings, including
+        # the needed policy ones.
+        all_params = (
```
Comment (simonsays1980) [inline rllib/algorithms/sac/torch/sac_torch_rl_module.py:144]: Nice!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Great PR with a big achievement. Multi-GPU on SAC is awesome!

File: rllib/BUILD (hunk 2 above)
Comment (sven1977) [reply rllib/BUILD:391]: Good point. We need to get rid of this confusion some time soon. Note that these are the command line options, not directly translatable to Algo config properties:
Here:
--num-cpus are the ray provided CPUs for the entire cluster.
--num-gpus are the number of Learner workers; note that if no GPUs are available, `--num-gpus` still sets the number of Learner workers, but then each worker gets one CPU (instead of 1 GPU). :|

File: rllib/BUILD (hunk 1 above)
Comment (sven1977) [reply rllib/BUILD:358]: For IMPALA/APPO, we should add a validation that these should never be run with a local Learner, b/c these are async algos that suffer tremendously from having the Learner not-async. Will add this check/error in a separate PR ...
